src/vm/eventpipe.cpp

```diff
@@ -236,27 +240,27 @@ void EventPipe::Shutdown()
     FlushProcessWriteBuffers();
 
     // Free resources.
-    delete(pConfig);
-    delete(pBufferManager);
-    delete(s_pEventSource);
+    delete pConfig;
+    delete pBufferManager;
+    delete s_pEventSource;
     s_pEventSource = NULL;
-    delete(s_pOutputPath);
+    delete[] s_pOutputPath;
     s_pOutputPath = NULL;
 
     // On Windows, this is just a pointer to the return value from
     // GetCommandLineW(), so don't attempt to free it.
 #ifdef FEATURE_PAL
-    delete[](s_pCommandLine);
+    delete[] s_pCommandLine;
     s_pCommandLine = NULL;
 #endif
 }
 
 EventPipeSessionID EventPipe::Enable(
     LPCWSTR strOutputPath,
-    unsigned int circularBufferSizeInMB,
-    EventPipeProviderConfiguration *pProviders,
-    int numProviders,
-    UINT64 multiFileTraceLengthInSeconds)
+    uint32_t circularBufferSizeInMB,
+    uint64_t profilerSamplingRateInNanoseconds,
+    const EventPipeSessionProviderList &providers,
```

Comment:
Sorry, I did not explain properly and in details at the top. Unless I missed something this is not quite the reason why I moved the code around.

`struct EventPipeProviderConfiguration` [here](https://github.com/dotnet/coreclr/blob/78570a239101f69200cfceab5e7527ca8cc312b8/src/vm/eventpipe.h#L358) is used in the *managed-to-unmanaged* calls. The owner of those strings is the managed code. We cannot deallocate them in the destructor for this struct. Thus, I think `EventPipeProviderConfiguration` should be only used on the *managed-to-unmanaged* calls.

On the other hand, there is another type `class EventPipeSessionProvider` [here](https://github.com/dotnet/coreclr/blob/78570a239101f69200cfceab5e7527ca8cc312b8/src/vm/eventpipesession.h#L144) that we can use. This type has a real and clear ownership of the data it holds.

Both struct and class hold the same data/fields/layout, and we actually create the latter in a later part of the code, so I just moved it up.

Reply:
Arguably, I do believe the "EventPipe" code works, but it does need some refactoring. There is too much coupling between the different types and the current "file drop" implementation.

Reply:
> Arguably, I do believe the "EventPipe" code works, but it does need some refactoring

IMO this part of the change has not simplified the previous state. In general you want to minimize coupling between components by choosing the smallest surface area that accomplishes your goals. Exposing EventPipeSessionProvider + EventPipeSessionProviderList + EventPipeProviderConfiguration is a larger surface area than necessary.

> We cannot deallocate them in the destructor for this struct. Thus, I think EventPipeProviderConfiguration should be only used on the managed-to-unmanaged calls.

I think that is an unecessarily restrictive definition. The caller needed to ensure that the string data which backs the pointers in EventPipeProviderConfiguration stays alive at least as long as the structure itself. That happens to occur naturally in the p/invoke case and didn't occur naturally in your parser. However I am suggesting that if you simplify the protocol it will simplify your parser, and once you simplify the parser it will be pretty easy to use EventPipeProviderConfiguration as-is because you won't want it to delete the strings. IMO that would be the refactoring that really simplifies the overall code. I can write up a quick example of what I think that could look like.

Reply:
OK, you convinced me... I'll change it back and simplify the parser.